Add AzureFunctionsEndToEnd with prototype implementation

captainsafia opened the pull request:
This PR introduces an AzureFunctionsEndToEnd sample application with prototype implementations of certain APIs:

- APIs that currently target `IHostApplicationBuilder` have been reimplemented to target `HostBuilder`. This should be addressed on Azure Functions supports `IHostApplicationBuilder`.
- A custom `AzureFunctionsProjectResource` has been included that launches the `func host` as an executable resource. This is a temporary implementation until we support modeling a Functions application as a project resource. These APIs are formally proposed in https://github.com/dotnet/aspire/issues/5412. Support for mocking the Functions app as a project resource can be added by updating the mock API definitions as follows:

```csharp
public class AzureFunctionsProjectResource(string name) 
    : ProjectResource(name), IResourceWithEnvironment, IResourceWithArgs, IResourceWithServiceDiscovery { }

public static class AzureFunctionsProjectResourceExtensions
  public static IResourceBuilder<FunctionAppResource> AddAzureFunctionsProject<TProject>(this IDistributedApplicationBuilder builder, string name) where TProject : IProjectMetadata, new()
    {
        var projectDirectory = Path.GetDirectoryName(new TProject().ProjectPath)!;
        var resource = new AzureFunctionsProjectResource(name);
        return builder.AddResource(resource)
            .WithAnnotation(new TProject())
            .WithHttpEndpoint(name: "http", env: "AZFUNCHOSTPORT")
            .WithEnvironment("FUNCTIONS_WORKER_RUNTIME", "dotnet-isolated")
            .WithEnvironment("OTEL_DOTNET_EXPERIMENTAL_OTLP_EMIT_EXCEPTION_LOG_ATTRIBUTES", "true")
            .WithEnvironment("OTEL_DOTNET_EXPERIMENTAL_OTLP_EMIT_EVENT_LOG_ATTRIBUTES", "true")
            .WithEnvironment("OTEL_DOTNET_EXPERIMENTAL_OTLP_RETRY", "in_memory")
            .WithEnvironment("ASPNETCORE_FORWARDEDHEADERS_ENABLED", "true")
            .WithEnvironment("FUNCTIONS_WORKER_RUNTIME", "dotnet-isolated")
            .WithOtlpExporter()
            .WithManifestPublishingCallback(async (context) => {
                context.Writer.WriteString("type", "function.v0");
                context.Writer.WriteString("path", context.GetManifestRelativePath(new TProject().ProjectPath));
                await context.WriteEnvironmentVariablesAsync(resource);
                context.Writer.WriteStartObject("bindings");
                foreach (var s in new string[]{ "http", "https"}) {
                    context.Writer.WriteStartObject(s);
                    context.Writer.WriteString("scheme", s);
                    context.Writer.WriteString("protocol", "tcp");
                    context.Writer.WriteString("transport", "http");
                    context.Writer.WriteBoolean("external", true);
                    context.Writer.WriteEndObject();
                }
                context.Writer.WriteEndObject();
            });
    }
}
```

And making the following changes to `AzureFunctionsEndToEnd.Functions.csproj`:

```xml
  <PropertyGroup>
    <RunCommand>func</RunCommand>
    <RunArguments>start --no-build --csharp --verbose --worker-runtime dotnetIsolated --port $(AZFUNCHOSTPORT)</RunArguments>
     <RunWorkingDirectory>bin/output</RunWorkingDirectory>
  </PropertyGroup>
```

- A custom `WithReference` implementation is implemented to support mapping resource strings as needed until https://github.com/dotnet/aspire/issues/5408 is implemented.
 ###### Microsoft Reviewers: [Open in CodeFlow](https://microsoft.github.io/open-pr/?codeflow=https://github.com/dotnet/aspire/pull/5418)

Fixes #5412 
Fixes #5408

captainsafia committed:
Add AzureFunctionsEndToEnd with prototype implementation

captainsafia committed:
Updat EventHub extension package to latest mirrored version

radical commented:
A test should be added for this too, see https://github.com/dotnet/aspire/pull/5316/ for an example.

davidfowl commented:
I'm not sure we want to test this sample yet. This is part of bootstrapping the azure functions work.

eerhardt commented on playground/AzureFunctionsEndToEnd/AzureFunctionsEndToEnd.ApiService/Program.cs:
```diff
@@ -0,0 +1,55 @@
+
+var app = builder.Build();
+
+app.MapGet("/publish/asq", async (QueueServiceClient client, CancellationToken cancellationToken) =>
```
What do you think about simplifying this example, and just putting these as buttons on the web front end?

That way someone only needs to reason about the web app, and the functions app.

For example, right now in the AppHost Program.cs, the webfrontend has a reference to apiService, but it doesn't appear to use/call it.

eerhardt replied:
> I think if we wanted the sample to be realistic

These "playground apps" aren't necessarily realistic and aren't intended to be samples. They are just intended for us to be able to try things out ourselves. Keeping them simple is a virtue so we don't have to maintain more than we need.

captainsafia replied:
We ended up simplifying it the other way around by removing the web UI from this sample app altogether. I think it's important to keep the API -> queue -> Azure Functions flow so this feels like a good landing place.

eerhardt commented on playground/AzureFunctionsEndToEnd/AzureFunctionsEndToEnd.Functions/HostBuilderExtensions.cs:
```diff
@@ -0,0 +1,61 @@
+                    //.AddGrpcClientInstrumentation()
+                        .AddHttpClientInstrumentation();
+                });
+            // services.AddOpenTelemetryExporters(context.Configuration);
```
Where's this from?

eerhardt commented on playground/AzureFunctionsEndToEnd/AzureFunctionsEndToEnd.Functions/HostBuilderExtensions.cs:
```diff
@@ -0,0 +1,61 @@
+                .WithTracing(tracing =>
+                {
+                    tracing.AddAspNetCoreInstrumentation()
+                    // Uncomment the following line to enable gRPC instrumentation (requires the OpenTelemetry.Instrumentation.GrpcNetClient package)
```
I think we can remove this commented out code. No need to put it in here if we aren't using gRPC.

eerhardt commented on playground/AzureFunctionsEndToEnd/AzureFunctionsEndToEnd.Functions/Program.cs:
```diff
@@ -0,0 +1,17 @@
+    {
+        services.AddOpenTelemetry()
+            .UseFunctionsWorkerDefaults()
+            .UseOtlpExporter();
```
How come this is here and not in `.AddServiceDefaults`?

davidfowl replied:
IHostBuilder vs IHostApplicationBuilder

eerhardt replied:
AddServiceDefaults (called above) is also implemented in terms of IHostBuilder. It is in the `HostBuilderExtensions.cs` file.

eerhardt commented on playground/AzureFunctionsEndToEnd/AzureFunctionsEndToEnd.Functions/AzureFunctionsEndToEnd.Functions.csproj:
```diff
@@ -0,0 +1,37 @@
+    <None Update="host.json">
+      <CopyToOutputDirectory>PreserveNewest</CopyToOutputDirectory>
+    </None>
+    <None Update="local.settings.json">
```
This file doesn't appear to be in your change. Is that intentional?

captainsafia replied:
Yeah, we can actually nuke this.

I believe this is autogenerated by the Azure Functions templates. The `local.settings.json` file is used to drive configuration options to the Core Tools (like whether or not this is a dotnet worker for example). Since we configure these ourselves when call the Core Tools CLI directly from the Aspire executable resource we don't really need it.

davidfowl replied:
Is that representative of a real functions project though? Or should we add this file?

captainsafia replied:
> Is that representative of a real functions project though? Or should we add this file?

The default Functions templates put this file in the `.gitignore` as well ([ref](https://github.com/Azure/azure-functions-templates/blob/f27baf3eb1aeb284739721d63c9c68059e3f27a6/Functions.Templates/ProjectTemplate_v4.x/CSharp-Isolated/_gitignore#L4-L5)) so I think it is fine to omit it. We'll want to make sure all the settings in it work correctly once we move to the `dotnet run` model though.

eerhardt committed:
Fix RandomString implementation

davidfowl committed:
Hack to make the emulators work

eerhardt commented on src/Aspire.Hosting.Azure.Storage/AzureStorageResource.cs:
```diff
@@ -38,6 +38,13 @@ public class AzureStorageResource(string name, Action<ResourceModuleConstruct> c
+    /// </summary>
+    /// <returns></returns>
+    public ReferenceExpression GetEmulatorConnectionString() =>
+        ReferenceExpression.Create($"{AzureStorageEmulatorConnectionString.Create(blobPort: EmulatorBlobEndpoint.Port, queuePort: EmulatorQueueEndpoint.Port, tablePort: EmulatorTableEndpoint.Port)}");
```
@davidfowl - should this throw if `!IsEmulator`?

davidfowl replied:
Yep

davidfowl committed:
More cleanup

davidfowl committed:
Primary constructors

davidfowl committed:
More cleanup

davidfowl committed:
Added Aspire.Hosting.Azure.Functions

davidfowl committed:
Experiment with specifying the host storage

davidfowl committed:
Separate host storage from the app storage

davidfowl committed:
First cut of IResourceWithAzureFunctionsConfig and removed WithReference
- We now scan all of the env variables and resolve references instead.

davidfowl committed:
Tested with managed identity and worked around issues
- Blob triggers require queues
- There's a conflict with the connection string section

davidfowl committed:
Implement the functions interface on storage as well

captainsafia commented on src/Aspire.Hosting.Azure.Functions/AzureFunctionsProjectResourceExtensions.cs:
```diff
@@ -0,0 +1,150 @@
+        var resource = new AzureFunctionsProjectResource(name, "func", projectDirectory);
+
+        // Add the default storage resource if it doesn't already exist.
+        var storage = builder.Resources.OfType<AzureStorageResource>().FirstOrDefault(r => r.Name == "azure-functions-default-storage");
```
Are we assuming that if a user wants to configure their own `AzureWebJobsStorage` they must configure it with this resource name for us to pick it up?

That seems like a sensible constraint to me but wanted to confirm that's what's goin on here.

davidfowl replied:
Nope, they just need to use `WithHostStorage` to do so. This default is only effective if they don't call that method (we keep this auto injected default)

davidfowl replied:
We can't avoid needing a name and a storage resource.

eerhardt replied:
I'm saying - instead of hard-coding a name `"azure-functions-default-storage"`. Would it be better if we created an internal Annotation (or similar) that only we could add to "mark" the AzureStorageResource as the default one? That way we don't rely on the Name, which is something that a user can manipulate.

davidfowl replied:
I don't understand how that would work. This resource needs to exist at some point and at that point it needs to have a name.

eerhardt replied:
> and at that point it needs to have a name.

That's fine, we can use some name like `"azure-functions-default-storage"` to give it a name.

I'm talking about how we recognize which AzureStorageResource is the default one. Basing that check on the name doesn't seem great. Imagine if a user wrote code like:

```C#
builder.AddAzureStorage("azure-functions-default-storage");

builder.AddAzureFunctionsProject<Projects.MyFuncs>("MyFuncs");
```

if they did that, this code would pick up the storage account they added on line 1. But instead we really want them to say:

```C#
var defaultStorage = builder.AddAzureStorage("azure-functions-default-storage");

builder.AddAzureFunctionsProject<Projects.MyFuncs>("MyFuncs")
    .WithHostStorage(defaultStorage);
```

davidfowl replied:
Ah, you want us to mark the default one. That's fine. If they did the above it would explode saying that they are multiple resources with the same name. Our best bet is to pick a name something that has a low likelyhood of collision

captainsafia replied:
I don't think the collisions are super likely to happen but I added this to https://github.com/dotnet/aspire/issues/5563 to track in a follow-up.

captainsafia commented on src/Aspire.Hosting.Azure.Functions/AzureFunctionsProjectResourceExtensions.cs:
```diff
@@ -0,0 +1,150 @@
+
+                    // Before the resource starts, we want to apply the azure functions specific environment variables.
+                    // we look at all of the environment variables and apply the configuration for any resources that implement IResourceWithAzureFunctionsConfig.
+                    item.Annotations.Add(new EnvironmentCallbackAnnotation(static context =>
```
Nice!

With this model, we could update more integrations packages to implement the `IResourceWithAzureFunctionsConfig` interface and they would light up in the Azure Functions resource without any additional updates on the `Aspire.Hosting.Azure.Functions` package.

captainsafia commented on src/Aspire.Hosting.Azure.Functions/AzureFunctionsProjectResourceExtensions.cs:
```diff
@@ -0,0 +1,150 @@
+                            }
+                        }
+
+                        // REVIEW: We need to remove the existing values before adding the new ones as there's a conflict with the connection strings.
```
Is the conflict with the connection strings just the fact that the Funtions host expects connection strings in one format and the Aspire-injected resources don't but they both share the same resource names?

davidfowl replied:
Needs to be investigated. Something in the functions or extensions codebase attempts to look at the `ConnectionStrings__{connectionName}` and throws because it's looking for a specific template format (Key1=Value1;Key2=Value2)

captainsafia replied:
Tracked as a follow-up in https://github.com/dotnet/aspire/issues/5563.

captainsafia committed:
Simulate modeling Azure Functions as project resources

davidfowl committed:
Make it work

davidfowl committed:
Remove the port from launch settings.

davidfowl committed:
Remove other interfaces from functions resource

davidfowl committed:
Simplify functions sample
- Delete the web app

captainsafia committed:
Add smoke test for launching Functions app

captainsafia committed:
Always run inner functions build

captainsafia committed:
Wire up package version for Aspire.Hosting.Azure.Functions

captainsafia committed:
Fix AppHost reference for test project

captainsafia committed:
Move AzureFunctionsEndToEnd tests to project specific tests

captainsafia committed:
Add Requires attributes to AzureFunctions tests

captainsafia committed:
Skip AzureFunctionsTest on Helix until core tools is installed

eerhardt commented on src/Aspire.Hosting.Azure.Functions/Aspire.Hosting.Azure.Functions.csproj:
```diff
@@ -0,0 +1,19 @@
+    <TargetFramework>$(NetCurrent)</TargetFramework>
+    <IsPackable>true</IsPackable>
+    <PackageTags>aspire hosting azure functions</PackageTags>
+    <Description>Azure fuinctions resource types for .NET Aspire.</Description>
```
```suggestion
    <Description>Azure functions resource types for .NET Aspire.</Description>
```

radical committed:
Merge remote-tracking branch 'origin/main' into add-azfunctions-prototype

radical committed:
Add deps on azure functions cli on helix

radical committed:
fix helix payload definition

radical committed:
Fix playground tests build

radical committed:
Fix cli payload for helix

radical committed:
Set destinationPath on helix payload, the default is the root directory

radical committed:
Fix typo

radical committed:
another round - fix path

radical commented:
`fail: AzureFunctionsEndToEnd.AppHost.Resources.funcapp[0] 1: 2024-09-05T07:09:43.6820000Z Can't determine Project to build. Expected 1 .csproj or .fsproj but found 2`

radical committed:
Merge remote-tracking branch 'origin/main' into add-azfunctions-prototype

radical committed:
simplify payload spec

radical commented:
Also, the eventhubs emulator is failing in the playground test with https://github.com/dotnet/aspire/issues/5371 .

davidfowl commented:
> Also, the eventhubs emulator is failing in the playground test with https://github.com/dotnet/aspire/issues/5371 .

Do you see this locally or only on the CI?

radical commented:
> > > Also, the eventhubs emulator is failing in the playground test with #5371 .
> > 
> > 
> > Do you see this locally or only on the CI?
> 
> Also on codespaces . But not locally on macOS.

Could it be because of some need for Docker desktop? More logs would help.

radical commented:
/azp run

davidfowl commented:
> Could it be because of some need for Docker desktop? More logs would help.

Yep:

- https://github.com/dotnet/aspire/pull/4192

davidfowl commented:
Same as https://github.com/dotnet/aspire/issues/5097

radical commented:
Another to add to https://github.com/dotnet/aspire/issues/5325 then :|

davidfowl commented:
This ups the urgency of this feature 😄. We'll see if we can work around the need for docker desktop in aspire. Do you recommend we disable the event hubs in the application to unblock this PR?

eerhardt commented:
> This ups the urgency of this feature 😄. We'll see if we can work around the need for docker desktop in aspire. Do you recommend we disable the event hubs in the application to unblock this PR?

Another option is to get docker desktop installed on our test machines.

davidfowl commented:
> Another option is to get docker desktop installed on our test machines.

Docker desktop shouldn't be required, we have a real product gap.

eerhardt commented:
> Docker desktop shouldn't be required, we have a real product gap.

Agreed. But that doesn't mean we shouldn't fully test this feature until the product gap is addressed.

We've known about this gap for a few months now. I guess I don't understand why "This ups the urgency of this feature". Any time a container needs to talk to another container, and Docker desktop isn't installed, it doesn't work today. This is just another "+1".

davidfowl commented:
If installing docker desktop is quick then I agree 😄

captainsafia committed:
Update docstrings and README for Aspire.Hosting.Azure.Functions

captainsafia committed:
Appease the linter

captainsafia committed:
Conditionally skip EventHubs emulation in CI

captainsafia commented:
> `fail: AzureFunctionsEndToEnd.AppHost.Resources.funcapp[0] 1: 2024-09-05T07:09:43.6820000Z Can't determine Project to build. Expected 1 .csproj or .fsproj but found 2`

There's ongoing discussion about this in https://github.com/Azure/azure-functions-core-tools/issues/3594 and a potential bug fix in https://github.com/Azure/azure-functions-core-tools/pull/3773 although it hasn't been released yet... :/

captainsafia committed:
Skip AzureFunctionsTest

radical commented:
Would you mind opening an issue to track the issue on our side, so we can enable the test when it is resolved, and use `[ActiveIssue("https://github.com/dotnet/aspire/issues/NNNN", typeof(PlatformDetection), nameof(PlatformDetection.IsRunningOnCI))]` on the test?

captainsafia commented:
> Would you mind opening an issue to track the issue on our side, so we can enable the test when it is resolved, and use `[ActiveIssue("https://github.com/dotnet/aspire/issues/NNNN", typeof(PlatformDetection), nameof(PlatformDetection.IsRunningOnCI))]` on the test?

Thanks for the pointer -- I'll update the PR with this.

captainsafia committed:
Use [ActiveIssue] to skip test in CI

radical commented on src/Aspire.Hosting.Azure.Functions/AzureFunctionsProjectResourceExtensions.cs:
```diff
@@ -0,0 +1,145 @@
+    /// </summary>
+    /// <typeparam name="TProject">The type of the project metadata, which must implement <see cref="IProjectMetadata"/> and have a parameterless constructor.</typeparam>
+    /// <param name="builder">The <see cref="IDistributedApplicationBuilder"/> to which the Azure Functions project will be added.</param>
+    /// <param name="name">The name to associated with the Azure Functions project. This name will be used for service discovery when referenced in a dependency.</param>
```
```suggestion
    /// <param name="name">The name to be associated with the Azure Functions project. This name will be used for service discovery when referenced in a dependency.</param>
```

radical commented on tests/helix/send-to-helix-inner.proj:
```diff
@@ -108,6 +111,17 @@
+
+    <HelixPreCommand Condition="'$(OS)' == 'Windows_NT'" Include="set PATH=%HELIX_CORRELATION_PAYLOAD%\azure-func-cli%3B%PATH%" />
+    <HelixPreCommand Condition="'$(OS)' != 'Windows_NT'" Include="export PATH=$HELIX_CORRELATION_PAYLOAD/azure-func-cli:$PATH" />
+
+    <HelixPreCommand Condition="'$(OS)' != 'Windows_NT'" Include="find $HELIX_CORRELATION_PAYLOAD/azure-func-cli -type d" />
```
This can be dropped now.

```suggestion
```

captainsafia replied:
I know it's effectively inert but I still think it's worthwhile to keep until the Core Tools update is unblocked...unless unnecessarily uploading this payloads to the Helix machine will cause unforeseen issues?

radical replied:
This will not block sending the payload to helix. I had added this line to check that the payload was being unpacked in the expected directory, for my own debugging.

radical commented on playground/Directory.Build.targets:
```diff
@@ -17,4 +20,8 @@
   <PropertyGroup Condition="'$(SkipDashboardProjectReference)' == 'true'">
     <DefineConstants>SKIP_DASHBOARD_REFERENCE;$(DefineConstants)</DefineConstants>
   </PropertyGroup>
+
+  <PropertyGroup Condition="'$(SkipEventHubsEmulation)' == 'true'">
+    <DefineConstants>SKIP_EVENTHUBS_EMULATION;$(DefineConstants)</DefineConstants>
+  </PropertyGroup>
```
```suggestion
  <PropertyGroup>
    <DefineConstants Condition="'$(SkipDashboardProjectReference)' == 'true'">SKIP_DASHBOARD_REFERENCE;$(DefineConstants)</DefineConstants>
    <DefineConstants Condition="'$(SkipEventHubsEmulation)' == 'true'">SKIP_EVENTHUBS_EMULATION;$(DefineConstants)</DefineConstants>
  </PropertyGroup>
```

captainsafia committed:
More feedback

captainsafia committed:
Update tests/helix/send-to-helix-inner.proj

Co-authored-by: Ankit Jain <radical@gmail.com>

radical commented:
Also, should this be fixed? [log](https://helixre107v0xdeko0k025g8.blob.core.windows.net/dotnet-aspire-refs-pull-5418-merge-c26ea5321b53455694/Aspire.Playground.Tests/1/console.9dcca55a.log?helixlogtype=result):
`/datadisks/disk1/work/A18408BE/w/B2030943/e/playground/AzureFunctionsEndToEnd/AzureFunctionsEndToEnd.Functions/AzureFunctionsEndToEnd.Functions.csproj : warning NU1603: Microsoft.Azure.Functions.Worker.Grpc 1.17.0 depends on Google.Protobuf (>= 3.25.2) but Google.Protobuf 3.25.2 was not found. An approximate best match of Google.Protobuf 3.25.3 was resolved. [/datadisks/disk1/work/A18408BE/w/B2030943/e/tests/Aspire.Playground.Tests/Aspire.Playground.Tests.csproj]`

captainsafia commented:
> Also, should this be fixed? [log](https://helixre107v0xdeko0k025g8.blob.core.windows.net/dotnet-aspire-refs-pull-5418-merge-c26ea5321b53455694/Aspire.Playground.Tests/1/console.9dcca55a.log?helixlogtype=result): `/datadisks/disk1/work/A18408BE/w/B2030943/e/playground/AzureFunctionsEndToEnd/AzureFunctionsEndToEnd.Functions/AzureFunctionsEndToEnd.Functions.csproj : warning NU1603: Microsoft.Azure.Functions.Worker.Grpc 1.17.0 depends on Google.Protobuf (>= 3.25.2) but Google.Protobuf 3.25.2 was not found. An approximate best match of Google.Protobuf 3.25.3 was resolved. [/datadisks/disk1/work/A18408BE/w/B2030943/e/tests/Aspire.Playground.Tests/Aspire.Playground.Tests.csproj]`

I just set a request to mirror Google.Protobuf@3.25.2.

radical committed:
Disable eventhubs on codespaces which doesn't have docker desktop

radical commented:
We'll need to resolve this issue before merging ([build](https://dev.azure.com/dnceng-public/public/_build/results?buildId=799520&view=logs&j=0bc77094-9fcd-5c38-f6e4-27d2ae131589&t=5138e247-88cb-5ba0-8eee-e2fdd127e7e8&l=638)):
`.dotnet\sdk\8.0.303\Microsoft.Common.CurrentVersion.targets(5270,5): error MSB3030: Could not copy the file "D:\a\_work\1\s\artifacts\obj\AzureFunctionsEndToEnd.Functions\Release\net8.0\extensions.json" because it was not found.`

captainsafia committed:
Update HTTP trigger to use ASP.NET-based trigger

captainsafia committed:
Merge branch 'main' into add-azfunctions-prototype

captainsafia commented:
/azp run

captainsafia commented:
/azp run

captainsafia commented:
/azp run

captainsafia committed:
Try 1: patch extensions.json copy bug

captainsafia committed:
Try 2: patch extensions.json copy bug

captainsafia committed:
Try 3: patch extensions.json copy bug

captainsafia committed:
Try only build AzureFunctionsEndToEnd project once in CI

captainsafia commented:
/azp run

captainsafia committed:
Bring back manual patch for extensions.json

captainsafia commented:
/azp run

captainsafia commented:
/azp run

captainsafia committed:
Enable include conditionally

captainsafia commented:
/azp run

captainsafia commented:
/azp run

captainsafia commented:
/azp run

captainsafia commented:
/azp run

captainsafia commented:
/azp run

dciborow commented:
standing by ready for to do some testing on our internal project once this is released. Please feel free to reach out - dciborow.

captainsafia commented:
/azp run

captainsafia commented:
> standing by ready for to do some testing on our internal project once this is released. Please feel free to reach out - dciborow.

Great to hear! I'll give the README another run through today to make sure we've captured all the limitations/setup instructions correctly. This is just the first part of the integration and there will be a few kinks to iron out. 😅

captainsafia committed:
Update README with workaround for build bug

captainsafia committed:
Fix markdown lint warnings

captainsafia committed:
Add note about lack of OTel support
